### PR TITLE
improved accuracy of gompertz distribution for small x

### LINKED ISSUE #4031: improve accuracy of scipy.stats.gompertz distribution for small x.
Currently the cdf-ppf round trip for small x return wrong values as shown here:
In [85]: gompertz.ppf(gompertz.cdf(1e-100,1),1)
Out[85]: 0.0

By replacing exp(x)-1 with expm1(x) and log(1+x) with log1p(x) one can improve the accuracy of the gompertz distribution (_logpdf, _cdf and _ppf)  to machine precision for small x.

Replacing  _logpdf, _cdf and _ppf with

    def _logpdf(self, x, c):
        return log(c) + x - c * expm1(x)

    def _cdf(self, x, c):
        return -expm1(-c * expm1(x))

    def _ppf(self, q, c):
        return log1p(-1.0 / c * log1p(-q))

one achieves 
 In [92]: gompertz.ppf(gompertz.cdf(1e-100,1),1)
Out[92]: 1e-100

### PR DESCRIPTION
**Pull request by rupak0577**
This pull request fixes #4031

### CODE REVIEW

**Comment by ev-br** [comment]
Thanks for the contribution!

First of all, the TravisCI failure is real: to use `expm1` and `log1p` you need to import them first. Either import them from `scipy.special` or use `special.expm1` etc.

By the way, it's a good idea to run tests locally before pushing: `$python runtests.py -s stats` would have caught it.

Which leads it to the second bit: this needs a test. An example fro gh-4031 would be a good one. I'd add it to `test_distributions.py`.

**Comment by rupak0577** [comment]
Thanks @ev-br. I will keep it in mind.

About the test:
I am new to scipy development and would like to know more on how to write the test for the gompertz distribution.

**Comment by ev-br** [comment]
A good regression test is   something which fails before your changes and does not fail when your patch is applied. In this case, the original issue, gh-4031, has an example which can be turned into a useful test case. 

Another example of a regression test: https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_distributions.py#L1783

File: scipy/stats/tests/test_distributions.py [hunk 1]
```diff
@@ -2159,6 +2159,10 @@ def test_infinite_input():
     assert_almost_equal(stats.skellam.sf(np.inf, 10, 11), 0)
     assert_almost_equal(stats.ncx2._cdf(np.inf, 8, 0.1), 1)
 
+def test_gompertz_accuracy():
+    # Regression test for gh-4031
+    p = stats.gompertz.ppf(stats.gompertz.cdf(1e-100,1),1)
+    assert_equal(p, 1e-100)
```
Comment (rgommers) [inline scipy/stats/tests/test_distributions.py]: This should be ``assert_allclose(p, 1e-100)`` - comparing floating point numbers for exact equality usually leads to trouble.

File: scipy/stats/_continuous_distns.py [hunk 2]
```diff
@@ -1932,13 +1932,13 @@ def _pdf(self, x, c):
...
-        return log(c) + x - c * (exp(x) - 1.)
+        return log(c) + x - c * special.expm1(x)
 
     def _cdf(self, x, c):
-        return 1.0-exp(-c*(exp(x)-1))
+        return -special.expm1(-c * special.expm1(x))
 
     def _ppf(self, q, c):
-        return log(1-1.0/c*log(1-q))
+        return special.log1p(-1.0/c*special.log1p(-q))
```
Comment (rgommers) [inline scipy/stats/_continuous_distns.py]: style: spaces around ``/`` would be nice

**Comment by rgommers** [comment]
@rupak0577 this has a merge conflict with scipy master. Do you know how to squash your commits and rebase? If so please do that, if not let us know and we'll help you.

**Comment by rgommers** [comment]
Also, it would be nice if you could make the commit message of the squashed commit in the format described at http://docs.scipy.org/doc/numpy-dev/dev/gitwash/development_workflow.html#writing-the-commit-message. Something like:

    ENH: improve accuracy of gompertz distribution for small x

**Comment by rupak0577** [comment]
@rgommers I would like to know the steps to rebase.

**Comment by maniteja123** [comment]
You can do ``git rebase -i master``. Then there will be a list of commits. Then select ``pick`` for the first commit and ``squash`` for the other commits. Then write the message for the squashed commit, according to the commit message style. Then push to the remote branch.

**Comment by rgommers** [comment]
I'd first squash and then rebase - fixing a merge conflict with a squash together may get annoying. This question gets asked often enough that I'm going to write it down here in detail and then make a PR. I can't find back previous descriptions.

Rebase:

    git checkout -b fix-4031-backup fix-4031  # not strictly needed, but good to have a back
    git checkout fix-4031   # go back to your working branch
    git fetch upstream    # git latest upstream master, so you can rebase on it
    git rebase upstream/master

This last command will stop when it hits a merge conflict.  Look at the message on your terminal; it's tell you which files have a conflict. You have to edit those files to make the merge (make sure to remove the ``<<<<``, ``=======`` and ``>>>>`` markers indicating conflicting sections). Save those files after editing, then:

    git add my_manually_merged_file_1.py
    git add my_manually_merged_file_2.py   # add all edited files like this
    git rebase --continue

This will finish the rebase. If there are multiple commits with a merge conflict, you may have to repeat the steps above to fix them.

Squashing commits:

    git rebase -i upstream/master   # re-arrange commits and add "squash" instead of "pick" where you want

Leaving the first commit as ``pick`` and all other as ``squash`` will get you one commit; you'll get dropped into the editor to update the commit message. You can delete all no longer needed messages and change it to ``ENH: improve accuracy of gompertz distribution for small x``.

If you first want to squash and then rebase, use ``gitk --all`` to find the last commit that's in both your branch and master, then use:

    git rebase -i last_common_commit_id

Note: alternative to looking at ``gitk`` is ``git merge-base --fork-point upstream/master last_common_commit_id``

**Comment by maniteja123** [comment]
@rgommers Thanks for the correct response, sorry for guiding the wrong way. I understood the situtation in wrong sense. I think it is a great idea to document this in the development workflow, which helps newcomers to contribution to a great extent.

**Comment by rupak0577** [comment]
@rgommers @maniteja123 Thanks for helping. I rebased but now it shows that my branch and origin/fix-4031 have diverged. Also, I created a new commit for the required changes but I am unable to squash the other commits.

**Comment by rgommers** [comment]
@rupak0577 yes it will show that they diverged - that's not a problem as long as your local branch looks fine. You then have to force-push it, which is the last step, forgot to describe that one.

    git push origin fix-4031 -f   # without -f Github will refuse to cooperate

Don't know why you can't squash all commits though. Can you push it without a squash?

**Comment by rgommers** [comment]
@maniteja123 no worries. Usually ``rebase -i`` works fine, just not always when combining it with a merge conflict fix.

**Comment by rupak0577** [comment]
@rgommers Pushed it without squashing. Thanks.

**Comment by rgommers** [comment]
OK squashed and merged in 8e3a683. Thanks @rupak0577

**Comment by rgommers** [comment]
One other thing: for the release notes we include a list of author names. Do you mind telling me (here or via email - ralf dot gommers at gmail.com) your last name? 

You can also set it via

    git config --global user.name "Your Name Comes Here"

as explained in http://docs.scipy.org/doc/numpy-dev/dev/gitwash/development_setup.html

**Comment by rupak0577** [comment]
@rgommers My full name Rupak Das.

**Comment by rgommers** [comment]
@rupak0577 thanks!

**Pull request closed without merging**